pkg/kubernetes/helper.go

```diff
@@ -61,13 +65,85 @@ func RemoveFinalizer(obj metav1.Object, toRemove ...string) {
 	obj.SetFinalizers(set.List())
 }
 
+func TryRemoveFinalizer(ctx context.Context, client ctrlruntimeclient.Client, obj ctrlruntimeclient.Object, finalizers ...string) error {
+	key := ctrlruntimeclient.ObjectKeyFromObject(obj)
+
+	err := retry.RetryOnConflict(retry.DefaultRetry, func() error {
+		// fetch the current state of the object
+		if err := client.Get(ctx, key, obj); err != nil {
+			// finalizer removal normally happens during object cleanup, so if
+			// the object is gone already, that is absolutely fine
+			if kerrors.IsNotFound(err) {
+				return nil
+			}
+			return err
+		}
+
+		// modify it
+		previous := sets.NewString(obj.GetFinalizers()...)
+		RemoveFinalizer(obj, finalizers...)
+		current := sets.NewString(obj.GetFinalizers()...)
+
+		// save some work
+		if previous.Equal(current) {
+			return nil
+		}
+
+		// update the object
+		return client.Update(ctx, obj)
```

Comment:
A lot of the code this replaced uses as `Patch()` instead of an `Update()`. Any reason we're going with an update here? Wouldn't a patch be more likely to succeed?